src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorSourceUpdater.cs

```diff
@@ -48,59 +56,332 @@ public async Task<IUpdaterResult> IsApplicableAsync(IUpgradeContext context, Imm
+                            documentText.Insert(startOffset, $"@{{ {updatedText} }}");
+                        }
+                    }
+                    else
+                    {
+                        // If the original text was completely removed, also search for implicit and explicit Razor expression syntax (@ or @()) so that it will be cleaned up, too
+                        if (string.IsNullOrWhiteSpace(updatedText))
+                        {
+                            var implicitExpression = $"@{originalText.Replace(";", string.Empty)}";
+                            var explicitExpression = $"@({originalText.Replace(";", string.Empty).Trim()})";
+                            documentText.Replace(implicitExpression, updatedText, startOffset, endOffset - startOffset);
+                            documentText.Replace(explicitExpression, updatedText, startOffset, endOffset - startOffset);
+                        }
+
+                        documentText.Replace(originalText, updatedText, startOffset, endOffset - startOffset);
+                    }
+                }
+
+                File.WriteAllText(replacementGroup.Key, documentText.ToString());
+            }
+        }
 
-            if (project is null)
+        // Removes leading and trailing portions of original and updated that are the same
+        private static void MinimizeReplacement(ref string original, ref string? updated)
+        {
+            if (updated is null)
             {
-                _logger.LogWarning("No project available.");
                 return;
             }
 
-            _logger.LogTrace("Running upgrade analyzers on Razor files in {ProjectName}", project.Name);
+            var index = 0;
+            while (index < original.Length && index < updated.Length && original[index] == updated[index])
+            {
+                index++;
+            }
+
+            var endIndex = 0;
+            while (endIndex > -original.Length && endIndex > -updated.Length && original[original.Length - 1 + endIndex] == updated[updated.Length - 1 + endIndex])
+            {
+                endIndex--;
+            }
+
+            original = original.Substring(index, Math.Max(0, original.Length + endIndex - index));
+            updated = updated.Substring(index, Math.Max(0, updated.Length + endIndex - index));
+        }
+
+        private static int GetLineOffset(RazorCodeDocument razorDoc, int startingLine)
+        {
+            var offset = 0;
+
+            for (var i = 1; i < startingLine && i <= razorDoc.Source.Lines.Count; i++)
+            {
+                // StreamSourceDoc.Lines is 0-based but line directives (as used in MappedSubText) are 1-based,
+                // so subtract one from i.
+                offset += razorDoc.Source.Lines.GetLineLength(i - 1);
+            }
+
+            return offset;
+        }
 
+        private static string GetGeneratedFilePath(RazorCodeDocument doc) => $"{doc.Source.FilePath}.cs";
```

Comment:
are the compiled razor files written to disk? if so, where?

Reply:
No, only in memory.